### ISSUE TITLE
How to fail the build?

### ISSUE THREAD

**Issue by CharliePoole** [cake-build/cake#573]
When tests fail, I detect the rc in a helper method and throw an exception. The task under which I'm running fails and no further tasks are executed.  Thisi s what I expect.

However, I also expect my Travis build to fail and that isn't happening. How can I make the cake.exe  and the build.sh fail in a way that Travis will notice it?

Note that since it's NUnit itself I'm testing, I can't use any features in the nunit or nunit 3 tools.

**Comment by CharliePoole** [comment]
BTW, I would have asked this on a forum rather than file an issue, but I can't find any info about  where questions should get asked and answered.

**Comment by devlead** [comment]
Normally most build servers fail depending on process return code, Cake changes return code on error / exception so that should fail the build.

We Build Cake with Cake on Travis for Linux and Mac and these are our settings
```yml
language: csharp
script:
  - ./build.sh
os:
  - linux
  - osx
cache:
  directories:
    - src/packages
    - tools
```
In our [build.sh](https://github.com/cake-build/cake/blob/develop/build.sh#L83) you'll find last line returns the return code of "last" process.
```bash
exit $?
```
Which is what passes along the information to the build server.

Perhaps this is what's missing from your build.sh?

**Comment by CharliePoole** [comment]
Do most people modify build.sh? If that's the intent, there's no problem. I've been following the rule "Don't modify it if somebody else maintains it." :-)

Does the same go for build.ps1? I've been tempted to modify that and held off for the same reason.

**Comment by CharliePoole** [comment]
So I added the exit $? to my otherwise unmodified build.sh. No joy. I take it that Cake.exe is still returning zero.

Here are some snippets of what I'm doing in my build.cake script:

```
Task("TestConsole")
  .IsDependentOn("Build")
  .Does(() => 
	{ 
		RunTest(NUNIT3_CONSOLE, BIN_DIR, CONSOLE_TESTS);
	});

...

void RunTest(FilePath exePath, DirectoryPath workingDir, string arguments)
{
	int rc = StartProcess(
	  MakeAbsolute(exePath), 
	  new ProcessSettings()
	  {
		  Arguments = arguments,
		  WorkingDirectory = workingDir
	  });
	  
	if (rc > 0)
	  throw new Exception(string.Format("{0} tests failed", rc));
	else if (rc < 0)
	  throw new Exception(string.Format("{0} returned rc = {1}", exePath, rc));
}
```

It's clear that Cake sees the exception, since it displays the message.  However, it appears to return zero itself as seen here: https://travis-ci.org/nunit/nunit/jobs/95706225

It's the return code from Cake.exe that I need to influence.

**Comment by CharliePoole** [comment]
I wondered if the exception might need to be thrown from within the Task, as seems to be implied by this page: http://cakebuild.net/docs/fundamentals/error-handling

However, after modifying the helper method to return the exit code and the Task as follows, I still see the same result. I added an echo to show $* and it's still zero.

```
Task("TestConsole")
  .IsDependentOn("Build")
  .Does(() => 
	{ 
		int rc = RunConsoleTest(NUNIT3_CONSOLE, BIN_DIR, CONSOLE_TESTS);

    if (rc > 0)
      throw new Exception(string.Format("{0} tests failed", rc));
    else if (rc < 0)
      throw new Exception(string.Format("Console returned rc = {0}", rc));
	});
```

**Comment by CharliePoole** [comment]
@devlead I examined your build script for Cake. Because you only use the Build target on Travis, it looks like the situation above actually can't arise.

**Comment by devlead** [comment]
@CharliePoole no most shouldn't have to modify the bootstrapper scripts, if they're lacking the report return code that should be fixed.
The PowerShell build.ps1 doesn't have this issue.

We aren't running tests on mono currently because xUnit issues under mono, as soon as they're solved well run them again.

Ah we don't throw when using StartProcess based on return code, as different tools may have different mechanisms of reporting failure. StartProcess is a very low level alias.

For running Nunit3 you should use one of the [Nunit3 aliases](http://cakebuild.net/dsl/nunit-v3), they understand NUnit parameters and results, and thereby will throw on failure, offer typed parameters and validation.

Unhandled exceptions will always change Cake return code.

**Comment by CharliePoole** [comment]
I haven't looked at @rprouse 's code, but I'm pretty sure the NUnit3 aliases are tools. That is, they will download the latest package and use it to run tests.

That's a classic problem of testing a test framework. You can never use the available tasks and tools because you're always ahead of them. When we used NAnt, we couldn't use the NUnit NAnt tasks. When we use MsBuild we can't use the contributed NUnit task. The same would appear to be the case for Cake: we have just built the version of nunit3-console that we want to use.

Additionally, we have to test under NUnitLite, where the test itself is the exe. So we basically need a way to run an arbitrary executable and capture it's return code. This seems like a very basic thing.

Back to your comments: I'm not saying you should throw based on the StartProcess return code. As you say, different executables will require different interpretation logic. My script is already interpreting the return code and throwing the exception when needed. The thrown exception is being handled by Cake.exe, which terminates immediately. But it terminates with a zero rc. How can I tell it to use a different return code?

**Comment by devlead** [comment]
@CharliePoole that makes sense, we've been thinking of adding possibility to override process arguments to tools, that would allow you to use new features with less overhead (discussed in issue #500 )
I've had Ideas of creating a generic  tool alias with basically just an toolpath, argus and expect lambda where you give an expression, so you get to write less pluming.

Regarding exit code Cake will return a not 0 return code on exception/error, is this the [build.sh](https://github.com/nunit/nunit/blob/issue-930/build.sh), if so could you try adding `exit $?` at end similar to [build.sh#L83](https://github.com/cake-build/cake/blob/develop/build.sh#L83).
It's included in our [recommended](http://cakebuild.net/docs/tutorials/setting-up-a-new-project) MacOSX [build.sh](http://cakebuild.net/bootstrapper/osx) and Generix posix [build.sh](https://raw.githubusercontent.com/cake-build/bootstrapper/master/res/scripts/build.sh) if we missed it in any place please raise an issue and we'll sort that out.

If it's nunit/nunit#1059 that's giving your working on and having these issues I could take a look and see if I can see any issue there beyond the suggested `exit $?`.

**Comment by gep13** [comment]
@CharliePoole in addition to @devlead comments, I just wanted to mention one other thing....

The NUnit alias that @rprouse created relies on the user/consumer adding a `#tool' line to the top of their build script (an example can be seen here: https://github.com/gep13/ChocolateyPackages/blob/master/gep13.DefaultBuild/Content/build.cake#L4) or by adding the tool to their packages.config file (an example can be seen here: https://github.com/gep13/ReSharperReports/blob/develop/Tools/packages.config#L5)

With either of these things in place, the alias within Cake will then look in the standard places to try to find the required `.exe` to do the work.  However, each tool derives from `ToolSettings` which includes a `ToolPath` property:

https://github.com/cake-build/cake/blob/develop/src/Cake.Core/Tooling/ToolSettings.cs#L14

as a result, as part of the build process, you can point the NUnit alias to the generated NUnit.exe which was created as part of your build process.

Now, as @devlead points out, if as part of the new build of NUnit, you are adding in additional command line switches, the pre-built alias will not be able to help you, but that is where the additional feature of https://github.com/cake-build/cake/issues/500 would come into play.

Hope that makes sense, and provides some help.

**Comment by rprouse** [comment]
@gep13 I didn't realize that you could set your own `ToolPath`. That would allow us to partially test the NUnit 3 task as we are testing NUnit itself.

You guys think of everything :smile: 

@CharliePoole, the build.sh file in your PR doesn't have the `exit $?` at the end and differs from the one linked to above. I believe that you fetched it using the instructions in the README?

```
curl -Lsfo build.sh http://cakebuild.net/bootstrapper/linux
```

I checked the URI and it is correct now, so not sure where the change came in.

On an aside, @devlead what are your problems with running your XUnit tests on Mono? Anything I can help with?

**Comment by gep13** [comment]
>@rprouse said...
@gep13 I didn't realize that you could set your own ToolPath. That would allow us to partially test the NUnit 3 task as we are testing NUnit itself.


@rprouse that is a relatively new addition to Cake, but it is one that now exists on all tools, as it is part of the base class.  That one is all thanks to @patriksvensson :-D

**Comment by CharliePoole** [comment]
@devlead @rprouse The code in nunit/nunit#1059 is definitely **not** what is being tested here. As previously commented, I set that branch aside to work on a branch that would not trigger four CI buiilds for every small change to the script. That branch is nunit/nunit/travis-issue-930.

However, my most recent comments and discussions on this issue have been from local code. To save reading back, the essence is:
* build.sh does have the return statement in it - I don't remember where I got my copy but I'll try to find it
* I locally echo the return from Cake.exe and have verified that it does not return anything but zero
* I'm throwing an exception with a message and not handling it in any way
* Cake handles the exception, displays the message and terminates with a zero rc

I'll put together a test script for this

**Comment by CharliePoole** [comment]
@gep13 That's indeed a great feature. I would probably add a pre-test to verify that the new build is actually being used, but that's just the kind of guy I am. :-) There's no issue with command-line options at this time - we just pass in the name of the test assembly.

I'll continue a few more iterations with the brute force approach before switching, however, because I'd like to see this cake issue resolved.

**Comment by devlead** [comment]
@CharliePoole are your scripts available publicly in some repo somewhere so I can test locally and in our CI env? As our 0.6.3 sadly broke posix builds high and low, we're clearly reporting return code for many people, and as the code's written I'm having hard time to see in which cases it wouldn't.
If there's an issue I would love to reproduce and see what's causing it so I can fix it or worst case find a temporary work around.

**Comment by CharliePoole** [comment]
I have to go to an appointment now but I'll put it out someplace when I get
back.

**Comment by CharliePoole** [comment]
Here's some negative evidence. Based on what I was seeing I thought this would fail, but it doesn't. 

**test.sh**
```
#!/bin/bash
mono ./tools/Cake/Cake.exe "test.cake"

echo "Return code was $?"
exit $?
```

**test.cake**
```
var target = Argument("target", "Test");

Task("Test")
	.Does(() =>
	{
		throw new Exception("Exception from Test task");
	});

RunTarget(target);
```

As you would expect, it displays

```
An error occured when executing task.
Error: Exception from Test task
Return code was 1
```

Except I didn't expect it, because the "same" code in my script is displaying an rc of 0! Clearly, it's not the "same" in some way. I'll see if I can reduce the script to the simplest possible example and then post it.

**Comment by CharliePoole** [comment]
OK, simplifying it didn't help at all. Take a look at this build:
    https://travis-ci.org/nunit/nunit/jobs/95869012 
and the corresponding branch on github:
    https://github.com/nunit/nunit/tree/ad607e287d7c388f82fbd816eb5933ff435ca258

BTW, it doesn't seem to matter whether I throw the exception from the RunTest helper method or from the task itself. I still see the same result.

**Comment by devlead** [comment]
@CharliePoole hi looking at the [build.sh](https://github.com/nunit/nunit/blob/ad607e287d7c388f82fbd816eb5933ff435ca258/build.sh#L63) script in the linked repo above it might be that your getting return value of the empty echo you have right after Cake.
As `$?`  context based it's a good idea to use it or store it immediately after executable is launched.

To store value the script could look something like this (mind I'm writing this on my phone so might sneak in an error or two, but hopefully you get the gist of it :))
```bash
# Start Cake
RETVAL=1
if $SHOW_VERSION; then
    mono $CAKE_EXE -version
elif $DRYRUN; then
    mono $CAKE_EXE $SCRIPT -verbosity=$VERBOSITY -configuration=$CONFIGURATION -target=$TARGET -dryrun
    RETVAL=$?
    echo
else
    mono $CAKE_EXE $SCRIPT -verbosity=$VERBOSITY -configuration=$CONFIGURATION -target=$TARGET
    RETVAL=$?
    echo
fi

echo "Last return code was $RETVAL"
exit $RETVAL
```

**Comment by CharliePoole** [comment]
I kind of doubt that since I added the echo late in the game. Also look
back in the comments at test.cake, which works as expected.

I'll try it later to confirm.

**Comment by CharliePoole** [comment]
Turns out that was it! I think it was failing even before I added the final echo because of the empty echoes right after executing the script.

I set up to use Cake via the cake-build/bootstrapper project, which doesn't include build.sh. I copied build.sh from the cake-build/example project. That build.sh includes the extra echoes and doesn't return an exit code.

Probably somebody should do a diff of all the different build.sh and build.ps1 scripts.

**Comment by gep13** [comment]
@CharliePoole Glad to hear that you have it all working now!  Let us know if you run into any other issues.

Can we close this issue now?

We now have an issue to pull together various sample scripts into what we are referring to as the Cake Contrib Repo, so I think all sample scripts will be reconciled into there going forward.

**Comment by CharliePoole** [comment]
I suggest fixing the one I used before closing the issue. It's incorrect for anyone and should match the final lines of your own shell script for the cake project.

You might also consider adding build.sh to the bootstrapper project.

**Comment by devlead** [comment]
The bootstraper has a [build.sh](https://github.com/cake-build/bootstrapper/blob/master/res/scripts/build.sh) which is a good starting point.

As @gep13 stated we're looking at consolidating and try to improve our documentation continuously, user feedback very important in that process.

**Comment by CharliePoole** [comment]
@devlead Well, that's kind of up to you guys. If it were me, I would not leave incorrect code for people to copy (in the example project) while working on a long-term solution. I suppose you could say it's technically not incorrect, since some folks might not care about the return code but it will be needed in any CI system I can think of.

I think it would help folks not working on windows or those who prefer to simply copy some starter scripts if both scripts were to be found in the root of the example project. At least to me, lacking the ability to actually run the bootstrapper project, that seemed like the right place to look.

**Comment by CharliePoole** [comment]
Looking back, the preceding sounds a bit harsh. Please chalk it up to something like two days I wasted because I copied the wrong build.sh file from the site.

**Comment by devlead** [comment]
Hi @CharliePoole!
cake-build/example#1 was merged yesterday and corrected the error in the sample [build.sh](https://github.com/cake-build/example/blob/master/build.sh).
I'm closing this issue now, If there are there any other questions raised you feel not answered in this issue please re-open or submit a new issue.
Thanks for your feedback.
/Mattias

**Comment by devlead** [comment]
Hi @CharliePoole & @rprouse,

Just FYI #588 closes #500 and adds the possibility to via delegate add or replace tool arguments will come with next release (now available as pre release on our MyGet feed).

All tools settings now have an
`Func<ProcessArgumentBuilder, ProcessArgumentBuilder> ArgumentCustomization`

This allows you to either totally replace arguments like this
```cake
NuGetInstall("MyNugetPackage", new NuGetInstallSettings {
    ArgumentCustomization = args=> "help"
    });
```
or reuse args i.e. same args against different tool versions by setting tool path
```cake
var myCustomPrevioslyDeclaredArg = new ProcessArgumentBuilder().Append("arg").AppendSecret("secret");
var toolPathsToTest = new [] {"./tools/nuget.exe", "./tools/nugetv2.exe", "./tools/nugetv3.exe"};
foreach(var toolPath in toolPathsToTest)
{
    Information("Testing: {0}", toolPath);
    NuGetInstall("MyNugetPackage", new NuGetInstallSettings {
        //Add New V3 option not yet avail
        ArgumentCustomization = args=>myCustomPrevioslyDeclaredArg,
        ToolPath = toolPath
        });
}
```
but also patch existing with new unknown commands
```cake
NuGetInstall("MyNugetPackage", new NuGetInstallSettings {
    //Add New V3 option not yet avail
    ArgumentCustomization = args=> args.AppendSwitchQuoted("-FallbackSource", " ", "https://www.myget.org/F/YOURFEED/api/v2")
    });
```

This reduces pluming code allot and together with ToolPath makes Cake very flexible.
